tests/unit/active/test_async_op_cuda.cc

```
@@ -0,0 +1,206 @@
+      ),
+      "cudaMemcpyAsync (stream1_)"
+    );
+
+    auto op1 = std::make_unique<messaging::AsyncOpCUDA>(stream1_);
+
+    checkCudaErrors(cudaStreamCreate(&stream2_), "cudaStreamCreate (stream2_)");
+    kernel<<<1, dataSize_, 0, stream2_>>>(dataDevicePointer2_, 8.0);
+    checkCudaErrors(
+      cudaMemcpyAsync(
+        dataHostPointer2_, dataDevicePointer2_, nBytes, cudaMemcpyDeviceToHost,
+        stream2_
+      ),
+      "cudaMemcpyAsync (stream2_)"
+    );
+
+    auto op2 = std::make_unique<messaging::AsyncOpCUDA>(
+      stream2_, [this] { done_ = true; }
+    );
+
+    // Register these async operations for polling; since these operations are
+    // enclosed in an epoch, they should inhibit the current epoch from
+    // terminating before they are done.
+    theMsg()->registerAsyncOp(std::move(op1));
+    theMsg()->registerAsyncOp(std::move(op2));
+  }
+
+  void check() {
+    for (size_t i = 0; i < dataSize_; ++i) {
+      EXPECT_DOUBLE_EQ(dataHostPointer1_[i], 4.0);
+      EXPECT_DOUBLE_EQ(dataHostPointer2_[i], 8.0);
+    }
+
+    EXPECT_TRUE(done_);
+  }
+
+private:
+  cudaStream_t stream1_;
+  cudaStream_t stream2_;
+
+  double* dataDevicePointer1_;
+  double* dataHostPointer1_;
+
+  double* dataDevicePointer2_;
+  double* dataHostPointer2_;
+
+  bool done_ = false;
+
+  size_t const dataSize_ = 32;
+};
+
+TEST_F(TestAsyncOp, test_async_op_cuda) {
+  int driverVer;
+  if (
+    (cudaDriverGetVersion(&driverVer) == cudaErrorInvalidValue) or
+    (driverVer == 0)) {
+    GTEST_SKIP()
+      << "Trying to run test_async_op_cuda but CUDA driver is not present!\n";
+  }
+
+  auto const this_node = theContext()->getNode();
+  auto p = theObjGroup()->makeCollective<CUDAGroup>("test_async_op_cuda");
+  auto ep = theTerm()->makeEpochRooted(term::UseDS{true});
+
+  // When this returns all the CUDA streams should be done
+  runInEpoch(ep, [p, this_node] {
+    p[this_node].send<MyMsg, &CUDAGroup::cudaHandler>();
+  });
+
+  cudaDeviceSynchronize();
```

Might this call hide potential failures? I think we want to run with checks without having synchronized, so we can have a chance of seeing incomplete data in the assertions